pkg/apis/apis/v1alpha1/types.go

```diff
@@ -0,0 +1,232 @@
+type ServiceBindingSpec struct {
+	// Services describe the APIs and implementations that are to be bound.
+	Services []ServiceSpec `json:"services,omitempty"`
+}
+
+// ServiceSpec describes one API and backing implementation.
+type ServiceSpec struct {
+	Native   *NativeServiceSpec   `json:"native,omitempty"`
+	Local    *LocalServiceSpec    `json:"local,omitempty"`
+	External *ExternalServiceSpec `json:"external,omitempty"`
+}
+
+// NativeServiceSpec describes an API and backing implementation that are provided by an actor on kcp.
+type NativeServiceSpec struct {
+	// Name of the ServiceExport that describes the API for this NativeServiceSpec.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Workspace of the ServiceExport that describes the API for this NativeServiceSpec.
+	//
+	// +kubebuilder:validation:Required
+	Workspace string `json:"workspace"`
+}
+
+// LocalServiceSpec describes an API and backing implementation that are provided by an actor in the
+// same Workspace as the ServiceBinding.
+type LocalServiceSpec struct {
+	// Name of the ServiceExport that describes the API for this LocalServiceSpec.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// ExternalServiceSpec describes an API and backing implementation that are provided by a third-party actor.
+type ExternalServiceSpec struct {
+	// URL at which this external service may be reached.
+	//
+	// +kubebuilder:validation:Format:uri
+	// +kubebuilder:validation:Required
+	URL string `json:"url"`
+}
+
+// ServiceBindingStatus records which schemas are bound.
+type ServiceBindingStatus struct {
+	// BoundServices records the state of binding individual services.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +patchStrategy=merge
+	// +patchMergeKey=name
+	BoundServices []BoundService `json:"boundServices,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// BoundService describes a bound ServiceExport.
+type BoundService struct {
+	// Name is the name of the ServiceExport which is bound.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name,omitempty"`
+	// BoundResourceSchemas record which resource schemas are bound for this ServiceExport.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=schema
+	// +patchStrategy=merge
+	// +patchMergeKey=schema
+	BoundResourceSchemas []ResourceSchemaReference `json:"boundResourceSchemas,omitempty" patchStrategy:"merge" patchMergeKey:"schema"`
+	// Current binding state of the ServiceExport.
+	//
+	// +optional
+	Conditions conditionsv1alpha1.Conditions `json:"conditions,omitempty"`
+}
+
+// ResourceSchemaReference records a resource, identified by (group, version, resource, schema).
+type ResourceSchemaReference struct {
+	metav1.GroupVersionResource `json:",inline"`
+
+	// Schema is the name of the bound ResourceSchema.
+	Schema string `json:"schema,omitempty"`
+	// Hash is
```

Comment:
.... ? we could even record the resourceVersion or generation, if those are simpler than an actual hash (do we care if the definition returns to a matching hash after changing to be different...?)

Reply:
wouldn't reference a RV or generation. It would make it hard to do shadow replication.